src/clj/harja/palvelin/palvelut/tehtavamaarat.clj

```diff
@@ -207,34 +208,54 @@
   (hae-tehtavahierarkia-maarineen db user {:urakka-id             urakka-id
                                            :hoitokauden-alkuvuosi hoitokauden-alkuvuosi}))
 
+(defn tallenna-sopimuksen-tehtavamaara [db user {:keys [urakka-id tehtava-id maara]}]
+  (let [urakkatyyppi (keyword (:tyyppi (first (urakat-q/hae-urakan-tyyppi db urakka-id))))
+        validit-tehtavat (hae-validit-tehtavat db)]
+  (oikeudet/vaadi-kirjoitusoikeus oikeudet/urakat-suunnittelu-tehtava-ja-maaraluettelo user urakka-id)
+
+  (when (empty?
+          (filter #(= tehtava-id
+                     (:tehtava-id %))
+            validit-tehtavat))
+    (throw (IllegalArgumentException. (str "Tehtävälle " tehtava-id " ei voi suunnitella määrätietoja."))))
+
+  (if-not (= urakkatyyppi :teiden-hoito)
+    (throw (IllegalArgumentException. (str "Urakka " urakka-id " on tyyppiä: " urakkatyyppi ". Urakkatyypissä ei suunnitella tehtävä- ja määräluettelon tietoja."))))
+  (q/tallenna-sopimuksen-tehtavamaara db user urakka-id tehtava-id maara)))
+
 (defrecord Tehtavamaarat []
   component/Lifecycle
   (start [this]
-    (doto (:http-palvelin this)
-      (julkaise-palvelu
-        :tehtavat
-        (fn [user tiedot]
-          (hae-tehtavat (:db this) user tiedot)))
-      (julkaise-palvelu
-        :tehtavahierarkia
-        (fn [user tiedot]
-          (hae-tehtavahierarkia (:db this) user tiedot)))
-      (julkaise-palvelu
-        :tehtavamaarat-hierarkiassa
-        (fn [user tiedot]
-          (hae-tehtavahierarkia-maarineen (:db this) user tiedot)))
-      (julkaise-palvelu
-        :tehtavamaarat
-        (fn [user tiedot]
-          (hae-tehtavamaarat (:db this) user tiedot)))
-      (julkaise-palvelu
-        :tallenna-tehtavamaarat
-        (fn [user tiedot]
-          (tallenna-tehtavamaarat (:db this) user tiedot)))
-      (julkaise-palvelu
-        :tehtavaryhmat-ja-toimenpiteet
-        (fn [user tiedot]
-          (tehtavaryhmat-ja-toimenpiteet (:db this) user tiedot))))
+    (let [db (:db this)]
```

Comment:
letin sijaan tämän :db (ja :http-palvelin) vois myös deffata suoraan tuosta (start [this]) destrukturoimalla